### PR TITLE
Adding column to Enrollment Record table

### PR DESCRIPTION
**Pull request by sudheshna-donthineni**
- Added columns `group_id` and `group_type` to Enrollment Record schema.
- `group_id` will contain the ID of the grouping (eg. group, batch, program)
- `group_type` will contain what kind of grouping. Will help in easier querying of groups.

- The ER controller, first checks if a record already exists. 
- It checks for a combo of `student_id`, `group_id`, `group_type`, `academic_year`. This means if a record of a student already exists with the same school and same grouping, that record needs to just be updated and no new record needs to be created.
- if there is no such record, then a new record is created.

### CODE REVIEW

File: lib/dbservice/schools/enrollment_record.ex [hunk 1]
```diff
@@ -34,17 +33,16 @@ defmodule Dbservice.Schools.EnrollmentRecord do
       :academic_year,
       :is_current,
       :board_medium,
-      :date_of_school_enrollment,
-      :date_of_group_enrollment,
-      :group_id
+      :group_id,
```
Comment (suryabulusu) [inline lib/dbservice/schools/enrollment_record.ex:36]: need to add date_of_enrollment here in cast

**Review by suryabulusu — COMMENTED** [review]
left a comment

**Review by suryabulusu — APPROVED** [review]

File: lib/dbservice/batches/batch.ex [hunk 2]
```diff
@@ -6,13 +6,15 @@ defmodule Dbservice.Batches.Batch do
...
   alias Dbservice.Groups.GroupType
+  alias Dbservice.Schools.EnrollmentRecord
 
   schema "batch" do
     field :name, :string
     field :contact_hours_per_week, :integer
 
     has_many :group_type, GroupType, foreign_key: :child_id, where: [type: "batch"]
     many_to_many :program, Program, join_through: "batch_program", on_replace: :delete
+    has_many :group_fk_id, EnrollmentRecord, foreign_key: :group_id
```
Comment (Bahugunajii) [inline lib/dbservice/batches/batch.ex:17]: Can you please help in understanding this association?

Comment (Bahugunajii) [inline lib/dbservice/batches/batch.ex:17]:
```suggestion
    has_many :group_fk, EnrollmentRecord, foreign_key: :group_id
```

File: lib/dbservice/groups/group.ex [hunk 3]
```diff
@@ -15,6 +16,7 @@ defmodule Dbservice.Groups.Group do
 
     has_many :program, Program
     has_many :group_type, GroupType, foreign_key: :child_id, where: [type: "group"]
+    has_many :group_fk_id, EnrollmentRecord, foreign_key: :group_id
```
Comment (Bahugunajii) [inline lib/dbservice/groups/group.ex:19]: Same here. Not able to understand the association.

File: lib/dbservice_web/controllers/enrollment_record_controller.ex [hunk 4]
```diff
@@ -74,15 +75,21 @@ defmodule DbserviceWeb.EnrollmentRecordController do
...
-      )
-      |> render("show.json", enrollment_record: enrollment_record)
+    existing_record =
+      index(conn, %{
+        "student_id" => params["student_id"],
+        "school_id" => params["school_id"],
+        "group_id" => params["group_id"],
+        "group_type" => params["group_type"],
+        "academic_year" => params["academic_year"]
+      })
```
Comment (Bahugunajii) [inline lib/dbservice_web/controllers/enrollment_record_controller.ex:85]: Or maybe we can directly pass the params?
Something like:
```suggestion
      index(conn, params)
```

**Review by Bahugunajii — CHANGES_REQUESTED** [review]

File: lib/dbservice/schools/enrollment_record.ex [hunk 5]
```diff
@@ -7,19 +7,18 @@ defmodule Dbservice.Schools.EnrollmentRecord do
...
   schema "enrollment_record" do
     field(:academic_year, :string)
     field(:grade, :string)
     field(:is_current, :boolean, default: false)
     field(:board_medium, :string)
-    field(:date_of_school_enrollment, :date)
-    field(:date_of_group_enrollment, :date)
+    field(:date_of_enrollment, :date)
+    field(:group_id, :integer)
+    field(:group_type, :string)
```
Comment (Bahugunajii) [inline lib/dbservice/schools/enrollment_record.ex:18]: What is the need of `group_type` here?

**Review by Bahugunajii — CHANGES_REQUESTED** [review]

**Review by Bahugunajii — CHANGES_REQUESTED** [review]

File: lib/dbservice_web/controllers/enrollment_record_controller.ex (hunk 4 above)
Comment (sudheshna-donthineni) [reply lib/dbservice_web/controllers/enrollment_record_controller.ex:85]: No, no, if we pass the whole params, all the parameters will be checked. We want only these specific ones to be checked.

File: lib/dbservice/batches/batch.ex (hunk 2 above)
Comment (sudheshna-donthineni) [reply lib/dbservice/batches/batch.ex:17]: Tagged in discord conversation.

File: lib/dbservice_web/controllers/enrollment_record_controller.ex (hunk 4 above)
Comment (Bahugunajii) [reply lib/dbservice_web/controllers/enrollment_record_controller.ex:85]: Index function will not return `existing_record` here. This seems an invalid approach (maybe you can also test it out to have more clarity). I would recommend to create a function in `school.exs` to handle this filter.

File: lib/dbservice/batches/batch.ex (hunk 2 above)
Comment (Bahugunajii) [reply lib/dbservice/batches/batch.ex:17]: Replied there!

File: lib/dbservice/schools/enrollment_record.ex (hunk 5 above)
Comment (sudheshna-donthineni) [reply lib/dbservice/schools/enrollment_record.ex:18]: For easier querying. We won't have to join the `group_type` table.

File: lib/dbservice/batches/batch.ex [hunk 6]
```diff
@@ -6,12 +6,14 @@ defmodule Dbservice.Batches.Batch do
...
   alias Dbservice.Programs.Program
   alias Dbservice.Groups.GroupType
+  alias Dbservice.Schools.EnrollmentRecord
 
   schema "batch" do
     field :name, :string
     field :contact_hours_per_week, :integer
 
     has_many :group_type, GroupType, foreign_key: :child_id, where: [type: "batch"]
+    has_many :group_fk, EnrollmentRecord, foreign_key: :group_id, where: [group_type: "batch"]
```
Comment (Bahugunajii) [inline lib/dbservice/batches/batch.ex:16]:
```suggestion
    has_many :enrollment_record, EnrollmentRecord, foreign_key: :group_id, where: [group_type: "batch"]
```

File: lib/dbservice/groups/group.ex [hunk 7]
```diff
@@ -15,6 +16,7 @@ defmodule Dbservice.Groups.Group do
 
     has_many :program, Program
     has_many :group_type, GroupType, foreign_key: :child_id, where: [type: "group"]
+    has_many :group_fk, EnrollmentRecord, foreign_key: :group_id, where: [group_type: "group"]
```
Comment (Bahugunajii) [inline lib/dbservice/groups/group.ex:19]:
```suggestion
    has_many :enrollment_record, EnrollmentRecord, foreign_key: :group_id, where: [group_type: "group"]
```

File: lib/dbservice/programs/program.ex [hunk 8]
```diff
@@ -23,6 +24,7 @@ defmodule Dbservice.Programs.Program do
     belongs_to :group, Group
     has_many :group_type, GroupType, foreign_key: :child_id, where: [type: "program"]
     many_to_many :batch, Batch, join_through: "batch_program", on_replace: :delete
+    has_many :group_fk, EnrollmentRecord, foreign_key: :group_id, where: [group_type: "program"]
```
Comment (Bahugunajii) [inline lib/dbservice/programs/program.ex:27]: Same here!

File: lib/dbservice/schools/enrollment_record.ex
Comment (Bahugunajii) [inline lib/dbservice/schools/enrollment_record.ex:1]: Any reasons of deleting this schema file?

**Review by Bahugunajii — CHANGES_REQUESTED** [review]

File: priv/repo/migrations/20240228055631_update_enrollment_record_table.exs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+defmodule Dbservice.Repo.Migrations.UpdateEnrollmentRecordTable do
+  use Ecto.Migration
+
+  def change do
+    alter table(:enrollment_record) do
+      remove :date_of_school_enrollment
+      remove :date_of_group_enrollment
+      add :date_of_enrollment, :date
+      add :group_type, :string
```
Comment (Bahugunajii) [inline priv/repo/migrations/20240228055631_update_enrollment_record_table.exs:9]: We should add `group_id` here also, right?

**Review by Bahugunajii — CHANGES_REQUESTED** [review]

File: lib/dbservice/enrollment_record/enrollment_record.ex [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  schema "enrollment_record" do
+    field(:academic_year, :string)
+    field(:grade, :string)
+    field(:is_current, :boolean, default: false)
+    field(:board_medium, :string)
+    field(:date_of_enrollment, :date)
+    field(:grouping_id, :integer)
+    field(:grouping_type, :string)
+
+    belongs_to(:student, Student)
```
Comment (Bahugunajii) [inline lib/dbservice/enrollment_record/enrollment_record.ex:19]:
```suggestion
    belongs_to(:student, Student)
    belongs_to(:batch, Batch)
    belongs_to(:group, Group)
    belongs_to(:program, Program)
```

File: lib/dbservice/enrollment_record/enrollment_record.ex [hunk 11]
```diff
@@ -0,0 +1,48 @@
+defmodule Dbservice.EnrollmentRecords.EnrollmentRecord do
+  @moduledoc false
+
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Dbservice.Utils.Util
+
+  alias Dbservice.Users.Student
```
Comment (Bahugunajii) [inline lib/dbservice/enrollment_record/enrollment_record.ex:8]:
```suggestion
  alias Dbservice.Users.Student
  alias Dbservice.Batches.Batch
  alias Dbservice.Groups.Group
  alias Dbservice.Programs.Program
```

**Review by Bahugunajii — CHANGES_REQUESTED** [review]

File: priv/repo/migrations/20240228055631_update_enrollment_record_table.exs (hunk 9 above)
Comment (sudheshna-donthineni) [reply priv/repo/migrations/20240228055631_update_enrollment_record_table.exs:9]: actually there already is a group_id in the table. But I do need to rename it.

**Review by Bahugunajii — APPROVED** [review]

**Pull request merged**